DownloadArtifacts errors in latest Cake 4.2.0 with 404 error

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched [issues](https://github.com/cake-build/cake/issues) to ensure it has not already been reported

### Cake runner

Cake Frosting

### Cake version

4.2.0

### Operating system

Linux

### Operating system architecture

x64

### CI Server

GitHub Actions

### What are you seeing?

see https://github.com/MonoGame/MonoGame/actions/runs/11728301127/job/32676762557?pr=8561

2024-11-07T19:29:45.1340265Z ========================================
2024-11-07T19:29:45.1341482Z DownloadArtifacts
2024-11-07T19:29:45.1342443Z ========================================
2024-11-07T19:29:45.4873409Z [41m[37;1mAn error occurred when executing task '[0m[41;1m[37;1mDownloadArtifacts[0m[41m[37;1m'.[0m
2024-11-07T19:29:45.4921275Z [41m[37;1mError: [0m[41;1m[37;1mArtifact service call GetSignedArtifactURL failed NotFound (404).[0m
2024-11-07T19:29:45.5987729Z ##[error]Process completed with exit code 255.

### What is expected?

The artifacts should be downloaded.

### Steps to Reproduce

This is the code we use

```
[TaskName("DownloadArtifacts")]
public sealed class DownloadArtifactsTask : AsyncFrostingTask<BuildContext>
{
    public override bool ShouldRun(BuildContext context) => context.BuildSystem().IsRunningOnGitHubActions;

    public override async Task RunAsync(BuildContext context)
    {
        context.CreateDirectory("nugets");
        foreach (var os in new[] { "windows", "macos", "linux" })
            await context.GitHubActions().Commands.DownloadArtifact($"nuget-{os}", "nugets");

        context.CreateDirectory("vsix");
        await context.GitHubActions().Commands.DownloadArtifact("MonoGame.Templates.VSExtension.vsix", "vsix");
    }
}
```

### Output log

2024-11-07T19:29:23.8793616Z Current runner version: '2.320.0'
2024-11-07T19:29:23.8817759Z ##[group]Operating System
2024-11-07T19:29:23.8818362Z Ubuntu
2024-11-07T19:29:23.8818835Z 22.04.5
2024-11-07T19:29:23.8819195Z LTS
2024-11-07T19:29:23.8819542Z ##[endgroup]
2024-11-07T19:29:23.8820000Z ##[group]Runner Image
2024-11-07T19:29:23.8820423Z Image: ubuntu-22.04
2024-11-07T19:29:23.8820823Z Version: 20241103.1.0
2024-11-07T19:29:23.8821900Z Included Software: https://github.com/actions/runner-images/blob/ubuntu22/20241103.1/images/ubuntu/Ubuntu2204-Readme.md
2024-11-07T19:29:23.8823366Z Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu22%2F20241103.1
2024-11-07T19:29:23.8824237Z ##[endgroup]
2024-11-07T19:29:23.8824688Z ##[group]Runner Image Provisioner
2024-11-07T19:29:23.8825174Z 2.0.384.1
2024-11-07T19:29:23.8825521Z ##[endgroup]
2024-11-07T19:29:23.8840065Z ##[group]GITHUB_TOKEN Permissions
2024-11-07T19:29:23.8841818Z Contents: write
2024-11-07T19:29:23.8842258Z Metadata: read
2024-11-07T19:29:23.8842970Z Packages: write
2024-11-07T19:29:23.8843466Z ##[endgroup]
2024-11-07T19:29:23.8846851Z Secret source: Actions
2024-11-07T19:29:23.8847464Z Prepare workflow directory
2024-11-07T19:29:23.9460552Z Prepare all required actions
2024-11-07T19:29:23.9616597Z Getting action download info
2024-11-07T19:29:24.1488138Z Download action repository 'actions/checkout@v4' (SHA:11bd71901bbe5b1630ceea73d27597364c9af683)
2024-11-07T19:29:24.2556203Z Download action repository 'actions/setup-dotnet@v4' (SHA:3e891b0cb619bf60e2c25674b222b8940e2c1c25)
2024-11-07T19:29:24.5820686Z Download action repository 'crazy-max/ghaction-github-runtime@v3' (SHA:b3a9207c0e1ef41f4cf215303c976869d0c2c1c4)
2024-11-07T19:29:24.8544162Z Download action repository 'ncipollo/release-action@v1' (SHA:2c591bcc8ecdcd2db72b97d6147f871fcd833ba5)
2024-11-07T19:29:25.1540575Z Complete job name: Deploy
2024-11-07T19:29:25.2348187Z ##[group]Run actions/checkout@v4
2024-11-07T19:29:25.2348659Z with:
2024-11-07T19:29:25.2348958Z   repository: MonoGame/MonoGame
2024-11-07T19:29:25.2349580Z   token: ***
2024-11-07T19:29:25.2349879Z   ssh-strict: true
2024-11-07T19:29:25.2350308Z   ssh-user: git
2024-11-07T19:29:25.2350630Z   persist-credentials: true
2024-11-07T19:29:25.2350986Z   clean: true
2024-11-07T19:29:25.2351292Z   sparse-checkout-cone-mode: true
2024-11-07T19:29:25.2351696Z   fetch-depth: 1
2024-11-07T19:29:25.2351980Z   fetch-tags: false
2024-11-07T19:29:25.2352283Z   show-progress: true
2024-11-07T19:29:25.2352602Z   lfs: false
2024-11-07T19:29:25.2352864Z   submodules: false
2024-11-07T19:29:25.2353170Z   set-safe-directory: true
2024-11-07T19:29:25.2353512Z ##[endgroup]
2024-11-07T19:29:25.4649180Z Syncing repository: MonoGame/MonoGame
2024-11-07T19:29:25.4650930Z ##[group]Getting Git version info
2024-11-07T19:29:25.4651663Z Working directory is '/home/runner/work/MonoGame/MonoGame'
2024-11-07T19:29:25.4652635Z [command]/usr/bin/git version
2024-11-07T19:29:25.4687861Z git version 2.47.0
2024-11-07T19:29:25.4712435Z ##[endgroup]
2024-11-07T19:29:25.4736703Z Temporarily overriding HOME='/home/runner/work/_temp/9072f744-05f7-4b5a-b2b5-52a229483980' before making global git config changes
2024-11-07T19:29:25.4739041Z Adding repository directory to the temporary git global config as a safe directory
2024-11-07T19:29:25.4742992Z [command]/usr/bin/git config --global --add safe.directory /home/runner/work/MonoGame/MonoGame
2024-11-07T19:29:25.4778363Z Deleting the contents of '/home/runner/work/MonoGame/MonoGame'
2024-11-07T19:29:25.4782078Z ##[group]Initializing the repository
2024-11-07T19:29:25.4786827Z [command]/usr/bin/git init /home/runner/work/MonoGame/MonoGame
2024-11-07T19:29:25.4884385Z hint: Using 'master' as the name for the initial branch. This default branch name
2024-11-07T19:29:25.4885496Z hint: is subject to change. To configure the initial branch name to use in all
2024-11-07T19:29:25.4886704Z hint: of your new repositories, which will suppress this warning, call:
2024-11-07T19:29:25.4887343Z hint:
2024-11-07T19:29:25.4887780Z hint: 	git config --global init.defaultBranch <name>
2024-11-07T19:29:25.4888635Z hint:
2024-11-07T19:29:25.4889156Z hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
2024-11-07T19:29:25.4890092Z hint: 'development'. The just-created branch can be renamed via this command:
2024-11-07T19:29:25.4890739Z hint:
2024-11-07T19:29:25.4891039Z hint: 	git branch -m <name>
2024-11-07T19:29:25.4912842Z Initialized empty Git repository in /home/runner/work/MonoGame/MonoGame/.git/
2024-11-07T19:29:25.4924196Z [command]/usr/bin/git remote add origin https://github.com/MonoGame/MonoGame
2024-11-07T19:29:25.4958518Z ##[endgroup]
2024-11-07T19:29:25.4959505Z ##[group]Disabling automatic garbage collection
2024-11-07T19:29:25.4963688Z [command]/usr/bin/git config --local gc.auto 0
2024-11-07T19:29:25.4993695Z ##[endgroup]
2024-11-07T19:29:25.4994701Z ##[group]Setting up auth
2024-11-07T19:29:25.5001515Z [command]/usr/bin/git config --local --name-only --get-regexp core\.sshCommand
2024-11-07T19:29:25.5033255Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :"
2024-11-07T19:29:25.5372023Z [command]/usr/bin/git config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2024-11-07T19:29:25.5403282Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :"
2024-11-07T19:29:25.5632521Z [command]/usr/bin/git config --local http.https://github.com/.extraheader AUTHORIZATION: basic ***
2024-11-07T19:29:25.5667754Z ##[endgroup]
2024-11-07T19:29:25.5668955Z ##[group]Fetching the repository
2024-11-07T19:29:25.5678180Z [command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +1b1c785ef9a9c3baf1388b0eb416fca80208df87:refs/remotes/origin/dev/dellis1972/cake-frosting
2024-11-07T19:29:26.8950470Z From https://github.com/MonoGame/MonoGame
2024-11-07T19:29:26.8951874Z  * [new ref]         1b1c785ef9a9c3baf1388b0eb416fca80208df87 -> origin/dev/dellis1972/cake-frosting
2024-11-07T19:29:26.8976703Z ##[endgroup]
2024-11-07T19:29:26.8977460Z ##[group]Determining the checkout info
2024-11-07T19:29:26.8979249Z ##[endgroup]
2024-11-07T19:29:26.8984317Z [command]/usr/bin/git sparse-checkout disable
2024-11-07T19:29:26.9026705Z [command]/usr/bin/git config --local --unset-all extensions.worktreeConfig
2024-11-07T19:29:26.9055499Z ##[group]Checking out the ref
2024-11-07T19:29:26.9060379Z [command]/usr/bin/git checkout --progress --force -B dev/dellis1972/cake-frosting refs/remotes/origin/dev/dellis1972/cake-frosting
2024-11-07T19:29:27.3280740Z Switched to a new branch 'dev/dellis1972/cake-frosting'
2024-11-07T19:29:27.3283444Z branch 'dev/dellis1972/cake-frosting' set up to track 'origin/dev/dellis1972/cake-frosting'.
2024-11-07T19:29:27.3307017Z ##[endgroup]
2024-11-07T19:29:27.3346182Z [command]/usr/bin/git log -1 --format=%H
2024-11-07T19:29:27.3369711Z 1b1c785ef9a9c3baf1388b0eb416fca80208df87
2024-11-07T19:29:27.3673501Z ##[group]Run actions/setup-dotnet@v4
2024-11-07T19:29:27.3673868Z with:
2024-11-07T19:29:27.3674104Z   dotnet-version: 8.0.x
2024-11-07T19:29:27.3674377Z   cache: false
2024-11-07T19:29:27.3674619Z ##[endgroup]
2024-11-07T19:29:27.5300655Z [command]/home/runner/work/_actions/actions/setup-dotnet/v4/externals/install-dotnet.sh --skip-non-versioned-files --runtime dotnet --channel LTS
2024-11-07T19:29:27.7777575Z dotnet-install: .NET Core Runtime with version '8.0.10' is already installed.
2024-11-07T19:29:27.7805965Z [command]/home/runner/work/_actions/actions/setup-dotnet/v4/externals/install-dotnet.sh --skip-non-versioned-files --channel 8.0
2024-11-07T19:29:28.0756346Z dotnet-install: .NET Core SDK with version '8.0.403' is already installed.
2024-11-07T19:29:28.0896354Z ##[group]Run crazy-max/ghaction-github-runtime@v3
2024-11-07T19:29:28.0896988Z with:
2024-11-07T19:29:28.0897691Z   github-token: ***
2024-11-07T19:29:28.0898074Z env:
2024-11-07T19:29:28.0898473Z   DOTNET_ROOT: /usr/share/dotnet
2024-11-07T19:29:28.0898834Z ##[endgroup]
2024-11-07T19:29:28.1266007Z ACTIONS_RUNNER_ACTION_ARCHIVE_CACHE=/opt/actionarchivecache
2024-11-07T19:29:28.1275241Z ACTIONS_RUNTIME_URL=https://pipelinesghubeus4.actions.githubusercontent.com/hQg2aSbEFP9s7Jip6HOjwAxh4oWhCQ2a4BMSxvy4ZfEiB9dgsE/
2024-11-07T19:29:28.1327770Z ACTIONS_RUNTIME_TOKEN=***
2024-11-07T19:29:28.1328719Z ACTIONS_CACHE_URL=https://acghubeus1.actions.githubusercontent.com/hQg2aSbEFP9s7Jip6HOjwAxh4oWhCQ2a4BMSxvy4ZfEiB9dgsE/
2024-11-07T19:29:28.1329917Z ACTIONS_RESULTS_URL=https://results-receiver.actions.githubusercontent.com/
2024-11-07T19:29:28.1401020Z ##[group]Run dotnet run --project build/Build.csproj -- --target=Deploy
2024-11-07T19:29:28.1401812Z [36;1mdotnet run --project build/Build.csproj -- --target=Deploy[0m
2024-11-07T19:29:28.1489312Z shell: /usr/bin/bash -e {0}
2024-11-07T19:29:28.1489971Z env:
2024-11-07T19:29:28.1490353Z   DOTNET_ROOT: /usr/share/dotnet
2024-11-07T19:29:28.1490873Z   ACTIONS_RUNNER_ACTION_ARCHIVE_CACHE: /opt/actionarchivecache
2024-11-07T19:29:28.1491932Z   ACTIONS_RUNTIME_URL: https://pipelinesghubeus4.actions.githubusercontent.com/hQg2aSbEFP9s7Jip6HOjwAxh4oWhCQ2a4BMSxvy4ZfEiB9dgsE/
2024-11-07T19:29:28.1532752Z   ACTIONS_RUNTIME_TOKEN: ***
2024-11-07T19:29:28.1533711Z   ACTIONS_CACHE_URL: https://acghubeus1.actions.githubusercontent.com/hQg2aSbEFP9s7Jip6HOjwAxh4oWhCQ2a4BMSxvy4ZfEiB9dgsE/
2024-11-07T19:29:28.1534762Z   ACTIONS_RESULTS_URL: https://results-receiver.actions.githubusercontent.com/
2024-11-07T19:29:28.1535498Z   GITHUB_TOKEN: ***
2024-11-07T19:29:28.1536052Z   NUGET_API_KEY: ***
2024-11-07T19:29:28.1536556Z   MARKETPLACE_PAT: ***
2024-11-07T19:29:28.1536979Z ##[endgroup]
2024-11-07T19:29:38.0199827Z /usr/share/dotnet/sdk/8.0.403/Microsoft.Common.CurrentVersion.targets(2412,5): warning MSB3246: Resolved file has a bad image, no metadata, or is otherwise inaccessible. PE image does not have metadata. PE image does not have metadata. [/home/runner/work/MonoGame/MonoGame/build/Build.csproj]
2024-11-07T19:29:45.0515620Z Version: 3.8.2.1570-develop
2024-11-07T19:29:45.0516505Z RepositoryUrl: https://github.com/MonoGame/MonoGame
2024-11-07T19:29:45.0517337Z BuildConfiguration: Release
2024-11-07T19:29:45.1339392Z 
2024-11-07T19:29:45.1340265Z ========================================
2024-11-07T19:29:45.1341482Z DownloadArtifacts
2024-11-07T19:29:45.1342443Z ========================================
2024-11-07T19:29:45.4873409Z [41m[37;1mAn error occurred when executing task '[0m[41;1m[37;1mDownloadArtifacts[0m[41m[37;1m'.[0m
2024-11-07T19:29:45.4921275Z [41m[37;1mError: [0m[41;1m[37;1mArtifact service call GetSignedArtifactURL failed NotFound (404).[0m
2024-11-07T19:29:45.5987729Z ##[error]Process completed with exit code 255.
2024-11-07T19:29:45.6110234Z Post job cleanup.
2024-11-07T19:29:45.7043285Z [command]/usr/bin/git version
2024-11-07T19:29:45.7081017Z git version 2.47.0
2024-11-07T19:29:45.7122530Z Temporarily overriding HOME='/home/runner/work/_temp/2675194b-04af-46b8-b5cf-45c3bc204a88' before making global git config changes
2024-11-07T19:29:45.7123826Z Adding repository directory to the temporary git global config as a safe directory
2024-11-07T19:29:45.7135748Z [command]/usr/bin/git config --global --add safe.directory /home/runner/work/MonoGame/MonoGame
2024-11-07T19:29:45.7169759Z [command]/usr/bin/git config --local --name-only --get-regexp core\.sshCommand
2024-11-07T19:29:45.7200768Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :"
2024-11-07T19:29:45.7487886Z [command]/usr/bin/git config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2024-11-07T19:29:45.7511842Z http.https://github.com/.extraheader
2024-11-07T19:29:45.7523687Z [command]/usr/bin/git config --local --unset-all http.https://github.com/.extraheader
2024-11-07T19:29:45.7556284Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :"
2024-11-07T19:29:45.8021865Z Cleaning up orphan processes
2024-11-07T19:29:45.8422109Z Terminate orphan process: pid (1929) (dotnet)

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -7,8 +7,19 @@ on:
       - develop
       - hotfix/*
 jobs:
+  prepare:
+    name: Prepare integration tests
+    runs-on: ubuntu-latest
+    steps:
+      - run: echo "Cake Integration Tests" > cake-integration-tests.txt
+      - uses: actions/upload-artifact@v4
+        with:
+          name: cake-integration-tests
+          path: cake-integration-tests.txt
+
   build:
     name: Build
+    needs: prepare
     runs-on: ${{ matrix.os }}
     strategy:
       fail-fast: false
@@ -34,7 +45,7 @@ jobs:
 
       - name: Run Cake script
         id: build-cake
-        uses: cake-build/cake-action@v1
+        uses: cake-build/cake-action@master
         with:
           target: Run-Integration-Tests
           cake-version: tool-manifest

--- a/src/Cake.Common.Tests/Fixtures/Build/GitHubActionsCommandsFixture.cs
+++ b/src/Cake.Common.Tests/Fixtures/Build/GitHubActionsCommandsFixture.cs
@@ -7,7 +7,6 @@ using System.Threading;
 using System.Threading.Tasks;
 using Cake.Common.Build.GitHubActions.Commands;
 using Cake.Common.Build.GitHubActions.Commands.Artifact;
-using Cake.Common.Build.GitHubActions.Data;
 using Cake.Common.Tests.Fakes;
 using Cake.Core;
 using Cake.Core.IO;
@@ -22,21 +21,45 @@ namespace Cake.Common.Tests.Fixtures.Build
         private const string ArtifactUrl = GitHubActionsInfoFixture.ActionResultsUrl + "twirp/github.actions.results.api.v1.ArtifactService/";
         private const string CreateArtifactUrl = ArtifactUrl + "CreateArtifact";
         private const string FinalizeArtifactUrl = ArtifactUrl + "FinalizeArtifact";
-        private const string GetSignedArtifactURLurl = ArtifactUrl + "GetSignedArtifactURL";
+        private const string GetSignedArtifactURLUrl = ArtifactUrl + "GetSignedArtifactURL";
+        private const string ListArtifactsUrl = ArtifactUrl + "ListArtifacts";
         private const string UploadFileUrl = "https://cake.build.net/actions-results/a9d82106-d5d5-4310-8f60-0bfac035cf02/workflow-job-run-1d849a45-2f30-5fbb-3226-b730a17a93af/artifacts/91e64594182918fa8012cdbf7d1a4f801fa0c35f485c3277268aad8e3f45377c.zip?sig=upload";
         private const string DownloadFileUrl = "https://cake.build.net/actions-results/a9d82106-d5d5-4310-8f60-0bfac035cf02/workflow-job-run-1d849a45-2f30-5fbb-3226-b730a17a93af/artifacts/91e64594182918fa8012cdbf7d1a4f801fa0c35f485c3277268aad8e3f45377c.zip?sig=download";
-        private const string CreateArtifactResponse = @"{
-    ""ok"": true,
-    ""signed_upload_url"": """ + UploadFileUrl + @"""
-}";
-        private const string FinalizeArtifactResponse = @"{
-    ""ok"": true,
-    ""artifact_id"": ""1991105334""
-}";
-        private const string GetSignedArtifactURLResponse = @"{ 
-    ""name"": ""artifact"",
-    ""signed_url"": """ + DownloadFileUrl + @"""
-}";
+        private const string CreateArtifactResponse =
+            $$"""
+            {
+                "ok": true,
+                "signed_upload_url": "{{UploadFileUrl}}"
+            }
+            """;
+        private const string FinalizeArtifactResponse =
+            """
+            {
+                "ok": true,
+                "artifact_id": "1991105334"
+            }
+            """;
+        private const string GetSignedArtifactURLResponse =
+            $$"""
+            { 
+                "name": "artifact",
+                "signed_url": "{{DownloadFileUrl}}"
+            }
+            """;
+        private const string ListArtifactsResponse =
+            $$"""
+            {
+              "artifacts": [
+                {
+                  "workflow_run_backend_id": "b9e28153-ca20-4b86-91dd-09e8f644efdf",
+                  "workflow_job_run_backend_id": "1d849a45-2f30-5fbb-3226-b730a17a93af",
+                  "database_id": "1",
+                  "name": "artifact",
+                  "created_at": "2024-11-09T21:53:00.7110204+00:00"
+                }
+              ]
+            }
+            """;
 
         private GitHubActionsInfoFixture GitHubActionsInfoFixture { get; }
         private ICakeEnvironment Environment { get; }
@@ -122,7 +145,7 @@ namespace Cake.Common.Tests.Fixtures.Build
                 // Get Signed Artifact Url
                 case
                 {
-                    RequestUri: { AbsoluteUri: GetSignedArtifactURLurl },
+                    RequestUri: { AbsoluteUri: GetSignedArtifactURLUrl },
                     Method: { Method: "POST" },
                 }:
                     {
@@ -187,6 +210,16 @@ namespace Cake.Common.Tests.Fixtures.Build
                 }:
                     {
                         return Ok();
+                    }
+
+                // List Artifacts
+                case
+                {
+                    RequestUri: { AbsoluteUri: ListArtifactsUrl },
+                    Method: { Method: "POST" }
+                }:
+                    {
+                        return Ok(new StringContent(ListArtifactsResponse));
                     }
 
                 // Download File

--- a/src/Cake.Common/Build/GitHubActions/Commands/Artifact/GitHubActionsArtifactService.cs
+++ b/src/Cake.Common/Build/GitHubActions/Commands/Artifact/GitHubActionsArtifactService.cs
@@ -30,16 +30,43 @@ namespace Cake.Common.Build.GitHubActions.Commands.Artifact
         private static readonly Uri CreateArtifactUrl = new Uri("CreateArtifact", UriKind.Relative);
         private static readonly Uri FinalizeArtifactUrl = new Uri("FinalizeArtifact", UriKind.Relative);
         private static readonly Uri GetSignedArtifactURLUrl = new Uri("GetSignedArtifactURL", UriKind.Relative);
+        private static readonly Uri ListArtifactsUrl = new Uri("ListArtifacts", UriKind.Relative);
 
         internal async Task DownloadArtifactFiles(
             string artifactName,
             DirectoryPath directoryPath)
         {
+            var listArtifactsResponse = await ListArtifacts(
+                artifactName);
+
+            if (listArtifactsResponse.Artifacts.FirstOrDefault(artifact => artifact.Name == artifactName)
+                is { WorkflowRunBackendId.Length: > 0 } and { WorkflowJobRunBackendId.Length: > 0 } artifact)
+            {
+                var signedArtifactURLResponse = await GetSignedArtifactURL(artifact.WorkflowRunBackendId, artifact.WorkflowJobRunBackendId, artifactName);
+
+                await DownloadArtifact(signedArtifactURLResponse.SignedUrl, directoryPath);
+            }
+            else
+            {
+                throw new CakeException($"Artifact {artifactName} not found.");
+            }
+        }
+
+        private async Task<ListArtifactsResponse> ListArtifacts(
+            string nameFilter = null,
+            long? idFilter = null)
+        {
             GetWorkflowBackendIds(out var workflowRunBackendId, out var workflowJobRunBackendId);
 
-            var (_, signedUrl) = await GetSignedArtifactURL(workflowRunBackendId, workflowJobRunBackendId, artifactName);
+            var listArtifactsRequest = new ListArtifactsRequest(
+                workflowRunBackendId,
+                workflowJobRunBackendId,
+                nameFilter,
+                idFilter);
 
-            await DownloadArtifact(signedUrl, directoryPath);
+            return await PostArtifactService<ListArtifactsRequest, ListArtifactsResponse>(
+                ListArtifactsUrl,
+                listArtifactsRequest);
         }
 
         private async Task DownloadArtifact(string signedUrl, DirectoryPath directoryPath)

--- a/src/Cake.Common/Build/GitHubActions/Commands/Artifact/ListArtifactsRequest.cs
+++ b/src/Cake.Common/Build/GitHubActions/Commands/Artifact/ListArtifactsRequest.cs
@@ -1,0 +1,20 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text.Json.Serialization;
+
+namespace Cake.Common.Build.GitHubActions.Commands.Artifact
+{
+#pragma warning disable SA1313 // Parameter names should begin with lower-case letter
+    internal record ListArtifactsRequest(
+        [property: JsonPropertyName("workflow_run_backend_id")]
+        string WorkflowRunBackendId,
+        [property: JsonPropertyName("workflow_job_run_backend_id")]
+        string WorkflowJobRunBackendId,
+        [property: JsonPropertyName("name_filter")]
+        string NameFilter = null,
+        [property: JsonPropertyName("id_filter")]
+        long? IdFilter = null);
+#pragma warning restore SA1313 // Parameter names should begin with lower-case letter
+}

--- a/src/Cake.Common/Build/GitHubActions/Commands/Artifact/ListArtifactsResponse.cs
+++ b/src/Cake.Common/Build/GitHubActions/Commands/Artifact/ListArtifactsResponse.cs
@@ -1,0 +1,27 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Text.Json.Serialization;
+
+namespace Cake.Common.Build.GitHubActions.Commands.Artifact
+{
+#pragma warning disable SA1313 // Parameter names should begin with lower-case letter
+    internal record ListArtifactsResponse(
+        [property: JsonPropertyName("artifacts")]
+        ListArtifactsResponse.MonolithArtifact[] Artifacts)
+    {
+        internal record MonolithArtifact(
+            [property: JsonPropertyName("workflow_run_backend_id")]
+            string WorkflowRunBackendId,
+            [property: JsonPropertyName("workflow_job_run_backend_id")]
+            string WorkflowJobRunBackendId,
+            [property: JsonPropertyName("database_id")]
+            string DatabaseId,
+            [property: JsonPropertyName("name")]
+            string Name,
+            [property: JsonPropertyName("created_at")]
+            DateTimeOffset CreatedAt);
+    }
+}

--- a/tests/integration/Cake.Common/Build/GitHubActions/GitHubActionsProvider.cake
+++ b/tests/integration/Cake.Common/Build/GitHubActions/GitHubActionsProvider.cake
@@ -137,6 +137,21 @@ Task("Cake.Common.Build.GitHubActionsProvider.Commands.DownloadArtifact")
         Assert.True(FileHashEquals(data.AssemblyPath, targetArtifactPath), $"{data.AssemblyPath.FullPath}=={targetArtifactPath.FullPath}");
 });
 
+Task("Cake.Common.Build.GitHubActionsProvider.Commands.DownloadArtifact.PreviousJob")
+    .Does(async () => {
+        // Given
+        var targetPath = Paths.Temp.Combine("./Cake.Common.Build.GitHubActionsProvider.Commands.DownloadArtifact.PreviousJob");
+        EnsureDirectoryExists(targetPath);
+        var targetArtifactPath = targetPath.CombineWithFilePath("cake-integration-tests.txt");
+
+        // When
+        await GitHubActions.Commands.DownloadArtifact("cake-integration-tests", targetPath);
+
+        // Then
+        Assert.True(System.IO.File.Exists(targetArtifactPath.FullPath), $"{targetArtifactPath.FullPath} Missing");
+        Assert.Equal("Cake Integration Tests\n", System.IO.File.ReadAllText(targetArtifactPath.FullPath));
+});
+
 Task("Cake.Common.Build.GitHubActionsProvider.Environment.Runner.Architecture")
     .Does(() => {
         // Given / When
@@ -193,7 +208,8 @@ if (GitHubActions.Environment.Runtime.IsRuntimeAvailable)
     gitHubActionsProviderTask
         .IsDependentOn("Cake.Common.Build.GitHubActionsProvider.Commands.UploadArtifact.File")
         .IsDependentOn("Cake.Common.Build.GitHubActionsProvider.Commands.UploadArtifact.Directory")
-        .IsDependentOn("Cake.Common.Build.GitHubActionsProvider.Commands.DownloadArtifact");
+        .IsDependentOn("Cake.Common.Build.GitHubActionsProvider.Commands.DownloadArtifact")
+        .IsDependentOn("Cake.Common.Build.GitHubActionsProvider.Commands.DownloadArtifact.PreviousJob");
 }
 
 public class GitHubActionsData

